internal/stacksapi/client.go

```diff
@@ -0,0 +1,300 @@
+		authHeader:  fmt.Sprintf("Token %s", apiToken),
+		retrierOpts: DefaultRetrierOptions,
+	}
+
+	for _, opt := range opts {
+		if err := opt(client); err != nil {
+			return nil, err
+		}
+	}
+
+	return client, nil
+}
+
+type StackAPIRequest struct {
+	*http.Request
+
+	retrier   *roko.Retrier
+	bodyBytes []byte
+}
+
+type RequestOption func(*StackAPIRequest) error
+
+// WithRetrier sets the retrier for the request, overriding the default retrier belonging to the [Client].
+func WithRetrier(retrier *roko.Retrier) RequestOption {
+	return func(r *StackAPIRequest) error {
+		r.retrier = retrier
+		return nil
+	}
+}
+
+// newRequest creates a new API request suitable for dispatch to the Buildkite Stacks API with the given context, method, path, and body.
+func (c *Client) newRequest(ctx context.Context, method, path string, body any, opts ...RequestOption) (*StackAPIRequest, error) {
+	fullURL := c.baseURL.ResolveReference(&url.URL{Path: path})
+
+	var bodyBytes []byte
+	if body != nil {
+		var err error
+		bodyBytes, err = json.Marshal(body)
+		if err != nil {
+			return nil, fmt.Errorf("failed to encode request body: %w", err)
+		}
+	}
+
+	req, err := http.NewRequestWithContext(ctx, method, fullURL.String(), bytes.NewReader(bodyBytes))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create request: %w", err)
+	}
+
+	req.Header.Set("User-Agent", c.userAgent)
+	req.Header.Set("Authorization", c.authHeader)
+	req.Header.Set("Content-Type", "application/json")
+
+	logger := c.logger.With(
+		"method", req.Method,
+		"url", req.URL,
+	)
+	if c.logHTTPPayloads {
+		logger = logger.With(
+			"request_headers", req.Header,
+			"request_body", string(bodyBytes),
+		)
+	}
+
+	logger.DebugContext(ctx, "created request")
+
+	stackReq := &StackAPIRequest{
+		Request:   req,
+		retrier:   roko.NewRetrier(c.retrierOpts...),
+		bodyBytes: bodyBytes,
+	}
+
+	for _, opt := range opts {
+		err := opt(stackReq)
+		if err != nil {
+			return nil, fmt.Errorf("failed to apply request option: %w", err)
+		}
+	}
+
+	return stackReq, nil
+}
+
+// do performs the given *StackAPIRequest and returns the HTTP response, unmarshalling the body of the HTTP response into
+// val, given it's a pointer to the expected response type.
+//
+// The do function will handle retries automatically based on the retrier options set in the request.
+func (c *Client) do(ctx context.Context, req *StackAPIRequest, val any) (*http.Response, error) {
+	logger := c.logger.With(
+		"method", req.Method,
+		"url", req.URL.String(),
+	)
+
+	resp, err := roko.DoFunc(ctx, req.retrier, func(r *roko.Retrier) (*http.Response, error) {
+		// Create a fresh body reader for each retry attempt
+		req.Body = io.NopCloser(bytes.NewReader(req.bodyBytes))
+
+		logger.DebugContext(ctx, "sending request")
+		resp, err := c.httpClient.Do(req.Request)
+		if err != nil {
+			return nil, err
+		}
+
+		err = checkResponse(resp)
+		if err != nil {
+			errResp := err.(*ErrorResponse)
```

Comment:
This line would panic if `err` wasn't an `*ErrorResponse`. Which could happen, because `checkResponse` returns a `fmt.Errorf(...)` if either the response body couldn't be read to completion, or the `json.Unmarshal` failed.

Reply:
yep, good shout